src/tainting/Dataflow_tainting.ml

```diff
@@ -1713,8 +1713,19 @@ let transfer :
                  * See [Taint_lval_env] for details. *)
                 Lval_env.add lval_env' lval taints
               else
-                (* Instruction returns safe data, remove taints from lval.
-                 * See [Taint_lval_env] for details. *)
+                (* The RHS returns no taint, but taint could propagate by
+                 * side-effect too. So, we check whether the taint assigned
+                 * to 'lval' has changed to determine whether we need to
+                 * clean 'lval' or not. *)
+                let lval_taints_changed =
+                  let lval_taints_before =
+                    Lval_env.dumb_find in' lval |> status_to_taints
+                  in
+                  let lval_taints_after =
+                    Lval_env.dumb_find lval_env' lval |> status_to_taints
+                  in
+                  not (Taints.equal lval_taints_before lval_taints_after)
+                in
```

Comment:
Is it possible for us to remove the taint from lval when lval is actually tainted?

```
while c:
  x = source()
```

Initially x is not tainted.

For the first visit, x will be tainted. It changed from clean -> tainted, so we keep it tainted.

For the second visit, x will be considered tainted again. It didn't change: tainted -> tainted, so we clean it.

And we may not reach a fixed point.

I may be misunderstanding everything though, since I still need to be more familiar with tainting code. Just checking.

Reply:
> For the second visit, x will be considered tainted again. It didn't change: tainted -> tainted, so we clean it.

The `x` is only cleaned when the RHS is clean, and in `x = source()` the RHS is _NOT_ clean. Note the new code is under the `else` associated with `if has_taints then`. Does that make sense?

Reply:
Makes sense!